File: src/main/java/com/tngtech/jgiven/Annotations.java
package com.tngtech.jgiven;

import com.google.common.collect.ImmutableSet;
import com.intellij.openapi.project.Project;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.PsiClass;
import com.intellij.psi.search.GlobalSearchScope;

import java.util.Arrays;
import java.util.Collection;
import java.util.Objects;
import java.util.stream.Collectors;

public class Annotations {
    private static final String EXPECTED_SCENARIO_STATE = "com.tngtech.jgiven.annotation.ExpectedScenarioState";
    private static final String PROVIDED_SCENARIO_STATE = "com.tngtech.jgiven.annotation.ProvidedScenarioState";
    private static final String SCENARIO_STATE = "com.tngtech.jgiven.annotation.ScenarioState";

    public static final Collection<String> JGIVEN_SCENARIO_STATE_ANNOTATION_NAMES = ImmutableSet.of(
            "ExpectedScenarioState",
            "ProvidedScenarioState",
            "ScenarioState");

    public static final Collection<String> JGIVEN_SCENARIO_STATE_CLASS_NAMES = Arrays.asList(
            EXPECTED_SCENARIO_STATE, PROVIDED_SCENARIO_STATE, SCENARIO_STATE
    );
    public static final Collection<String> PROVIDED_SCENARIO_STATE_CLASS_NAMES = Arrays.asList(
            PROVIDED_SCENARIO_STATE, SCENARIO_STATE
    );
    public static final Collection<String> EXPECTED_SCENARIO_STATE_CLASS_NAMES = Arrays.asList(
            EXPECTED_SCENARIO_STATE, SCENARIO_STATE
    );

    private static PsiClass findPsiClass(String fqClassname, JavaPsiFacade javaPsiFacade, Project project) {
        return javaPsiFacade.findClass(fqClassname, GlobalSearchScope.allScope(project));
    }

    public static Collection<PsiClass> getScenarioStateClasses(Project project) {
        return getClassesFor(project, JGIVEN_SCENARIO_STATE_CLASS_NAMES);
    }

    private static Collection<PsiClass> getClassesFor(Project project, Collection<String> classNames) {
        return classNames.stream()
                .map(a -> findPsiClass(a, JavaPsiFacade.getInstance(project), project))
                .filter(Objects::nonNull)
                .collect(Collectors.toSet());
    }
}

File: src/main/java/com/tngtech/jgiven/Annotations.kt
package com.tngtech.jgiven

import com.intellij.openapi.project.Project
import com.intellij.psi.JavaPsiFacade
import com.intellij.psi.PsiClass
import com.intellij.psi.search.GlobalSearchScope
import java.util.*

object Annotations {
    private const val EXPECTED_SCENARIO_STATE = "com.tngtech.jgiven.annotation.ExpectedScenarioState"
    private const val PROVIDED_SCENARIO_STATE = "com.tngtech.jgiven.annotation.ProvidedScenarioState"
    private const val SCENARIO_STATE = "com.tngtech.jgiven.annotation.ScenarioState"

    val JGIVEN_SCENARIO_STATE_CLASS_NAMES: Collection<String> = Arrays.asList(
            EXPECTED_SCENARIO_STATE, PROVIDED_SCENARIO_STATE, SCENARIO_STATE
    )
    val PROVIDED_SCENARIO_STATE_CLASS_NAMES: Collection<String> = Arrays.asList(
            PROVIDED_SCENARIO_STATE, SCENARIO_STATE
    )
    val EXPECTED_SCENARIO_STATE_CLASS_NAMES: Collection<String> = Arrays.asList(
            EXPECTED_SCENARIO_STATE, SCENARIO_STATE
    )

    private fun findPsiClass(fqClassname: String, javaPsiFacade: JavaPsiFacade, project: Project) =
            javaPsiFacade.findClass(fqClassname, GlobalSearchScope.allScope(project))

    fun getScenarioStateClasses(project: Project) =
            getClassesFor(project, JGIVEN_SCENARIO_STATE_CLASS_NAMES)

    private fun getClassesFor(project: Project, classNames: Collection<String>): Collection<PsiClass> =
            classNames
                    .mapNotNull { findPsiClass(it, JavaPsiFacade.getInstance(project), project) }
                    .toSet()
}
